ksqldb-engine/src/test/java/io/confluent/ksql/engine/QueryMonitorTest.java

```diff
@@ -0,0 +1,245 @@
+package io.confluent.ksql.engine;
+
+import com.google.common.base.Ticker;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import org.apache.kafka.streams.KafkaStreams;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import static org.apache.kafka.streams.KafkaStreams.State.ERROR;
+import static org.apache.kafka.streams.KafkaStreams.State.RUNNING;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class QueryMonitorTest {
```

Comment:
we should add a test case to ensure that the query monitor deals with terminated queries correctly.

Reply:
Done. `shouldNoRestartQueryThatWasManuallyTerminated`